beluga/include/beluga/sensor/likelihood_field_model3.hpp

```diff
@@ -0,0 +1,168 @@
+  /**
+   * Used to calculate the probability of the obstacle being hit.
+   */
+  double sigma_hit = 0.2;
+};
+
+/// Likelihood field sensor model for range finders.
+/**
+ * This model relies on a pre-computed likelihood map of the environment.
+ * It is less computationally intensive than the beluga::BeamSensorModel
+ * because no ray-tracing is required, and it can also provide better
+ * performance in environments with non-smooth occupation maps. See
+ * Probabilistic Robotics \cite thrun2005probabilistic, Chapter 6.4,
+ * for further reference.
+ *
+ * \note This class satisfies \ref SensorModelPage.
+ *
+ * \tparam OpenVDB grid type.
+ */
+template <typename GridT, typename PointCloud>
+class LikelihoodFieldModel3 {
+ public:
+  /// State type of a particle.
+  using state_type = Sophus::SE3d;
+
+  /// Weight type of the particle.
+  using weight_type = double;
+
+  /// Measurement type given by the interface.
+  using measurement_type = PointCloud;
+
+  /// Map representation type.
+  using map_type = GridT;
+
+  /// Parameter type that the constructor uses to configure the likelihood field model.
+  using param_type = LikelihoodFieldModel3Param;
+
+  /// Constructs a Likelihood3DFieldModel instance.
+  /**
+   * \param params Parameters to configure this instance.
+   *  See beluga::Likelihood3DFieldModelParam for details.
+   * \param grid Narrow band Level set grid representing the static map that the sensor model
+   *  uses to compute a likelihood field for lidar hits and compute importance weights
+   *  for particle states.
+   *  Currently only supports OpenVDB Level sets.
+   */
+  explicit LikelihoodFieldModel3(const param_type& params, const map_type& grid)
+      : params_{params},
+        grid_{openvdb::gridPtrCast<map_type>(grid.deepCopyGrid())},
+        accessor_{grid_->getConstAccessor()},
+        transform_{grid_->transform()},
+        background_{grid_->background()},
+        two_squared_sigma_{2 * params.sigma_hit * params.sigma_hit},
+        amplitude_{params.z_hit / (params.sigma_hit * std::sqrt(2 * Sophus::Constants<double>::pi()))},
+        offset_{params.z_random / params.max_laser_distance} {
+    openvdb::initialize();
+    /// Pre-computed parameters
+    assert(two_squared_sigma_ > 0.0);
+    assert(amplitude_ > 0.0);
+  }
+
+  /// Returns a state weighting function conditioned on 3D lidar hits.
+  /**
+   * \param measurement 3D lidar measurement containing the hit points and the transform to the origin.
+   * \return a state weighting function satisfying \ref StateWeightingFunctionPage
+   *  and borrowing a reference to this sensor model (and thus their lifetime are bound).
+   */
+  [[nodiscard]] auto operator()(measurement_type&& measurement) const {
+    const size_t pointcloud_size = measurement.points().size();
+    // Transform each point from the sensor frame to the origin frame
+    auto transformed_points = ranges::views::all(measurement.points()) |
```

Comment:
nit: `ranges::views::all` should not be needed. The transform view will do the conversion.

```suggestion
    auto transformed_points = measurement.points() |
```

Reply:
Changed in 2aabb60e1d02fcf2464ebca96f3cdb47be230d3a